uarray/_uarray_dispatch.cxx

```diff
@@ -0,0 +1,491 @@
+}
+
+
+struct py_func_args { py_ref args, kwargs; };
+
+struct Function
+{
+	PyObject_HEAD
+	py_ref extractor_, replacer_;   // functions to handle dispatchables
+	py_ref backends_;               // function returning backends as iterable
+	py_ref def_args_, def_kwargs_;  // default arguments
+	py_ref def_impl_;               // default implementation
+	py_ref dict_;                   // __dict__
+
+	PyObject * call(PyObject * args, PyObject * kwargs);
+
+	py_func_args replace_dispatchables(
+		PyObject * backend, PyObject * args, PyObject * kwargs, PyObject * coerce);
+
+	py_ref canonicalize_args(PyObject * args);
+	py_ref canonicalize_kwargs(PyObject * kwargs);
+
+	static void dealloc(Function * self)
+		{
+			PyObject_GC_UnTrack(self);
+			self->~Function();
+			Py_TYPE(self)->tp_free(self);
+		}
+
+	static PyObject * new_(PyTypeObject * type, PyObject * args, PyObject * kwargs)
+		{
+			auto self = reinterpret_cast<Function *>(type->tp_alloc(type, 0));
+			if (self == nullptr)
+				return nullptr;
+
+			// Placement new
+			self = new (self) Function;
+			return reinterpret_cast<PyObject *>(self);
+		}
+
+	static int init(Function * self, PyObject * args, PyObject * /*kwargs*/)
+		{
+			PyObject * extractor, * replacer;
+			PyObject * backends;
+			PyObject * def_args, * def_kwargs;
+			PyObject * def_impl;
+
+			if (!PyArg_ParseTuple(
+				    args, "OOOO!O!O",
+				    &extractor,
+				    &replacer,
+				    &backends,
+				    &PyTuple_Type, &def_args,
+				    &PyDict_Type, &def_kwargs,
+				    &def_impl))
+			{
+				return -1;
+			}
+
+			if (!PyCallable_Check(extractor) || !PyCallable_Check(replacer))
+			{
+				PyErr_SetString(PyExc_TypeError,
+				                "Argument extractor and replacer must be callable");
+				return -1;
+			}
+
+			if (!PyCallable_Check(backends))
+			{
+				PyErr_SetString(PyExc_TypeError,
+				                "The backends argument must be callable");
+				return -1;
+			}
+
+			if (def_impl != Py_None && !PyCallable_Check(def_impl))
+			{
+				PyErr_SetString(PyExc_TypeError,
+				                "Default implementation must be Callable or None");
+				return -1;
+			}
+
+			self->extractor_ = py_ref::ref(extractor);
+			self->replacer_ = py_ref::ref(replacer);
+			self->backends_ = py_ref::ref(backends);
+			self->def_args_ = py_ref::ref(def_args);
+			self->def_kwargs_ = py_ref::ref(def_kwargs);
+			self->def_impl_ = py_ref::ref(def_impl);
+
+			return 0;
+		}
+};
+
+
+bool is_default(PyObject * value, PyObject * def)
+{
+	// TODO: richer comparison for builtin types? (if cheap)
+	return (value == def);
+}
+
+
+py_ref Function::canonicalize_args(PyObject * args)
+{
+	const auto arg_size = PyTuple_GET_SIZE(args);
+	const auto def_size = PyTuple_GET_SIZE(def_args_.get());
+
+	if (arg_size > def_size)
+		return py_ref::ref(args);
+
+	Py_ssize_t mismatch = 0;
+	for (Py_ssize_t i = arg_size - 1; i >= 0; --i)
+	{
+		auto val = PyTuple_GET_ITEM(args, i);
+		auto def = PyTuple_GET_ITEM(def_args_.get(), i);
```

Comment:
There might be a negative access here since `def_args_` isn't sliced properly.

Reply:
The check above guaruntees that `def_size >= arg_size`

Reply:
Hmm, it's a bug in any case, and should be fixed.

Reply:
Not sure I follow, what's a bug?

Reply:
Never mind, mind derp.

Reply:
Shouldn't this be offset by `len(args)`?

Reply:
Missed your latest comment, still can't see what the issue is though. 
I check the pairs as if from `zip(args, def_args_)` starting from the end `i = len(args)-1` through to and including the first argument `i = 0`.

Reply:
My mind kept going back and forth on the indices required.